### ISSUE TITLE
Type safe clients with document references

### ISSUE THREAD

**Issue by acanturgut** [bijoutrouvaille/fireward#24]
Hi, 

I couldn't see to how to define DocumentReference as a type. Any idea about how to do that? 
```
type User = {
    discountCodes?: string[], 
    company?: DocumentReference
}
```

**Comment by bijoutrouvaille** [comment]
@acanturgut I am having a hard time coming up with a use case for that. Why not just store the path string?

**Comment by MrAlek** [comment]
I'd like to re-visit this issue. Sure document reference are really just syntactic sugar for a path string but Firestore, including the official client SDKs, have a DocumentReference type on which you can do operations specific to references and not just any string. Just like Timestamp is just syntactic sugar for a number type.

In the longer run, references could also be extended with types for the document the reference points to. In our own Flutter (Dart) codebase, we have created a TypedDocumentReference<T> which encodes the information of which type to parse a document snapshot to down the line. That would be very useful for Fireward's Typescript generators as well.

Example:
```
type Name = { first: string, last?: string }
type User = { name: Name | string, employedAt: documentReference<Company> }
type Company = { name: string }
```

**Comment by bijoutrouvaille** [comment]
Thanks @MrAlek for the follow-up. Just to clarify, the `timestamp` type is a primitive type in the Firestore rules language. And this is what Fireward does for TS: it maps rules types to TS types. I know there's a [Path](https://firebase.google.com/docs/reference/rules/rules.Path) type in the rules, but I don't know if it has a primitive correlate in rules and if it can be validated against. If it did map to the document reference, it would make sense to add that to Firestore generator.

But in any case, thank you for reporting on your usage. Fireward grows from input like this.

**Comment by MrAlek** [comment]
@bijoutrouvaille thanks for the quick reply & clarification.

Personally, I find great value in using Fireward for compiling rules but I'd also love to expand on the case of generating type safe models for clients. Coming from GraphQL and the excellent Apollo framework, this is the biggest thing I miss with Firestore and it doesn't look like Google is picking up the ball here.

Since `DocumentReference` is a type being used in the Typescript client SDK it would be very useful having it in the model definition, even though it doesn't directly map to a primitive in the rules language. What do you think?

**Comment by bijoutrouvaille** [comment]
@MrAlek I would also find a type aware client generator very useful. 

It seems like a different idea from storing doc references though. A DocReference is not just a type but a class with undocumented internal properties. Codifying it into the data structure seems like making it into a minefield.

Note that Fireward does not allow unspecified properties, unlike typescript.

**Comment by bijoutrouvaille** [comment]
I'm reopening this issue for a broader discussion of automatic model storage / retrieval with type safe clients.

**Comment by MrAlek** [comment]
@bijoutrouvaille That's true (re: DocumentReference). Perhaps a typealias could be used instead. It would be easy for anyone to expand on using it to create instances of DocumentReference classes. `type Path = string`
